Objects/moduleobject.c

```diff
@@ -739,6 +739,29 @@ _PyModuleSpec_IsInitializing(PyObject *spec)
     return 0;
 }
 
+/* Check if the submodule name is in the "_uninitialized_submodules" attribute
+   of the module spec.
+   Clear the exception and return 0 if spec is NULL.
+ */
+int
+_PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
+{
+    if (spec != NULL) {
+        _Py_IDENTIFIER(_uninitialized_submodules);
+        PyObject *value = _PyObject_GetAttrId(spec,
+                                              &PyId__uninitialized_submodules);
+        if (value != NULL) {
+            int is_uninitialized = PySequence_Contains(value, name);
+            Py_DECREF(value);
+            if (is_uninitialized >= 0) {
+                return is_uninitialized;
+            }
+        }
+    }
+    PyErr_Clear();
+    return 0;
+}
```

Comment:
I prefer to avoid nesting the conditionals so every case is more clear:

```suggestion
{
    if (spec == NULL) {
         goto exit;
    }
      
    _Py_IDENTIFIER(_uninitialized_submodules);
    PyObject *value = _PyObject_GetAttrId(spec, &PyId__uninitialized_submodules);
     if (value == NULL) {
         goto error;
     }
     int is_uninitialized = PySequence_Contains(value, name);
     Py_DECREF(value);
     if (is_uninitialized == -1) {
         goto error
     }
     return is_uninitialized;
exit:
    PyErr_Clear();
    return 0;
}
```

Reply:
Also, I don't get why we need to clear the error if `spec == NULL`. That sounds a bit weird as an API.

Reply:
This code was almost a verbatim copy from `_PyModuleSpec_IsInitializing`, it was not clear to me why the error was cleared given that we actually return on `spec == NULL && PyErr_Occurred()` in `module_getattro`. Looking at the commit that introduced it (https://github.com/python/cpython/commit/3e429dcc242e48fa4cbb1a91cf7c416c37b97b4e), it makes sense, but this is no longer needed.
We could remove this from `_PyModuleSpec_IsUninitializedSubmodule` and then I could make other PR removing it from `_PyModuleSpec_IsInitializing` too.

Reply:
I would prefer to remove both here to keep then consistent